sourcetracker/_sourcetracker.py

```diff
@@ -9,80 +9,92 @@
+    ValueError
+        If `sources` and `sinks` passed and columns are not identical.
     '''
-    if not np.isreal(feature_table.dtypes).all():
-        raise TypeError('Feature table contains one or more columns which are '
-                        'not numeric type. This is likely due to boolean, '
-                        'string, or other data types being present. Data must '
-                        'contain exclusively real-valued integers or floats.')
-
-    if np.isnan(feature_table.values).any():
-        raise ValueError('One or more values in the feature table is a `nan` '
-                         'or `null` value. Data must contain exclusively real-'
-                         'valued integers or floats.')
-
-    if not (np.ceil(feature_table.values) == feature_table.values).all():
-        if apply_fractional_value_correction is True:
-            return pd.DataFrame(np.ceil(feature_table.values).astype(np.int64),
-                                index=feature_table.index,
-                                columns=feature_table.columns)
-        else:
-            raise ValueError('Non-integer data in the feature table is not '
-                             'being corrected by `check_and_correct_data`. '
-                             'Data leaving this function must be np.int64 '
-                             'type. Either correct the table or pass the '
-                             '`apply_fractional_value_correction` to '
-                             '`check_and_correct_data`.')
-    return feature_table.astype(np.int64)
+    if sinks is not None:
+        dfs = [sources, sinks]
+    else:
+        dfs = [sources]
+
+    for df in dfs:
+        # Because of this bug (https://github.com/numpy/numpy/issues/6114)
+        # we can't use e.g. np.isreal(df.dtypes).all(). Instead we use
+        # applymap. Based on:
+        # http://stackoverflow.com/questions/21771133/finding-non-numeric-rows-in-dataframe-in-pandas
+        if not df.applymap(np.isreal).values.all():
+            raise ValueError('A dataframe contains one or more values which '
+                             'are not numeric. Data must be exclusively '
+                             'positive integers.')
+        if np.isnan(df.values).any():
+            raise ValueError('A dataframe has `nan` or `null` values. Data '
+                             'must be exclusively positive integers.')
+        if (df.values < 0).any():
+            raise ValueError('A dataframe has a negative count. Data '
+                             'must be exclusively positive integers.')
+
+    if sinks is not None:
+        if not (sinks.columns == sources.columns).all():
+            raise ValueError('Dataframes do not contain identical (and '
+                             'identically ordered) columns. Columns must '
+                             'match exactly.')
+        return sources.astype(np.int32), sinks.astype(np.int32)
```

Comment:
Could we do these type casts in place (but still return the relevant objects) for both ``return`` cases? I'm worried about this blowing up memory.